rustdoc: Render HRTB correctly for bare functions

camelid opened the pull request:
The angle brackets were not rendered, so code like this:

    some_func: for<'a> fn(val: &'a i32) -> i32

would be rendered as:

    some_func: fn'a(val: &'a i32) -> i32

However, rendering with angle brackets is still invalid syntax:

    some_func: fn<'a>(val: &'a i32) -> i32

so now it renders correctly as:

    some_func: for<'a> fn(val: &'a i32) -> i32

-----

However, note that this code:

    some_trait: dyn for<'a> Trait<'a>

will still render as:

    some_trait: dyn Trait<'a>

which is not invalid syntax, but is still unclear. Unfortunately I think
it's hard to fix that case because there isn't enough information in the
`rustdoc::clean::Type` that this code operates on. Perhaps that case can
be fixed in a later PR.

r? @jyn514

camelid commented on src/test/rustdoc/for-lifetime.rs:
```diff
@@ -0,0 +1,14 @@
+    pub some_trait: dyn for<'a> Trait<'a>,
+}
+
+// @has foo/index.html '//span[@id="structfield.some_func"]' "some_func: fn<'a>(val: &'a i32) -> i32"
+// @has foo/index.html '//span[@id="structfield.some_trait"]' "some_trait: dyn Trait<'a>"
```
These tests are failing :/

I wish `@has` checks gave me something better than:

```
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
11: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/index.html '//span[@id="structfield.some_func"]' "some_func: fn<'a>(val: &'a i32) -> i32"
12: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/index.html '//span[@id="structfield.some_trait"]' "some_trait: dyn Trait<'a>"

Encountered 2 errors

------------------------------------------

info: generating a diff against nightly rustdoc
warning: no pager configured, falling back to `diff --color`
help: try configuring a git pager (e.g. `delta`) with `git config --global core.pager delta`
thread '[rustdoc] rustdoc/for-lifetime.rs' panicked at 'assertion failed: output.status.success() || output.status.code() == Some(1)', src/tools/compiletest/src/runtest.rs:2469:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    [rustdoc] rustdoc/for-lifetime.rs
```

@jyn514 Didn't you make a PR at some point to add nice diffs? It seems like it doesn't work on my machine...

jyn514 replied:
@camelid did you try following the advice? ;)

```
info: generating a diff against nightly rustdoc
warning: no pager configured, falling back to `diff --color`
help: try configuring a git pager (e.g. `delta`) with `git config --global core.pager delta`
```

camelid replied:
What's a "diff pager"?

camelid replied:
I tried `git config --global core.pager less`, but it still doesn't work. It doesn't use the pager and just outputs a huge unstyled diff.

jyn514 replied:
Oh wait actually this is probably https://github.com/rust-lang/rust/pull/79370

jyn514 replied:
Try `brew install tidy` (or however you prefer to install programs)

camelid replied:
I already have tidy installed :)

camelid replied:
Also, I'm not getting that error.

jyn514 replied:
@camelid I copy-pasted that output from the error you posted in https://github.com/rust-lang/rust/pull/79991#discussion_r541808910. Are you getting a different error now?

jyn514 replied:
Also, it might help to rebase over https://github.com/rust-lang/rust/pull/79370.

camelid replied:
I was referring to the error in #79370 which is about `tidy`, not the error I posted which is about a pager.

jyn514 replied:
@camelid I added more tools for debugging this in https://github.com/rust-lang/rust/pull/79370, can you rebase over that and see what the new output is when you run the test?

camelid replied:
I rebased over `upstream/master` but I'm getting this error:

```
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
11: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/index.html '//span[@id="structfield.some_func"]' "some_func: fn<'a>(val: &'a i32) -> i32"
12: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/index.html '//span[@id="structfield.some_trait"]' "some_trait: dyn Trait<'a>"

Encountered 2 errors

------------------------------------------

info: generating a diff against nightly rustdoc
thread '[rustdoc] rustdoc/for-lifetime.rs' panicked at 'assertion failed: status.success() || status.code() == Some(1)', src/tools/compiletest/src/runtest.rs:2420:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    [rustdoc] rustdoc/for-lifetime.rs
```

Then I edited compiletest to show what the actual status code is:

```
thread '[rustdoc] rustdoc/for-lifetime.rs' panicked at 'HTML `tidy` returned unexpected status code Some(2)', src/tools/compiletest/src/runtest.rs:2420:21
```

Then I removed the assert and changed `diff --color` to `diff` since my version of diff doesn't seem to have a `--color` option, and it dumps a massive diff. What should I do?

jyn514 replied:
> Then I removed the assert and changed diff --color to diff since my version of diff doesn't seem to have a --color option,

This is a bug in compiletest, can you open an issue?

>  it dumps a massive diff

This is kind of just a limitation of how I wrote it :/ if you run `rustup default nightly` (or `rustup override set nightly`, which doesn't affect global state) it should at least have a *smaller* diff. It will never be exactly what you want to change though, that would require comparing to `git merge-base HEAD origin/master` which I don't have the infrastructure set up for, and I'm not quite sure how I'd add it.

camelid replied:
Opened #80037.

camelid replied:
It's also not paging the output, even though I set my pager to `less`.

camelid replied:
I managed to eventually generate a diff using this command:

```console
$ git diff --no-index --word-diff -- build/x86_64-apple-darwin/test/rustdoc/for-lifetime.nightly build/x86_64-apple-darwin/test/rustdoc/for-lifetime
```

```diff
<div class="docblock type-decl hidden-by-usual-hider"><pre class="rust struct">pub struct Foo {
    pub some_func: {+for<'a>+} <a class="primitive" [-href="https://doc.rust-lang.org/nightly/std/primitive.fn.html">fn</a>'a(val:-]{+href="https://doc.rust-lang.org/nightly/std/primitive.fn.html">fn</a>(val:+} &amp;'a <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.i32.html">i32</a>) -&gt; <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.i32.html">i32</a>,
    pub some_trait: dyn <a class="trait" href="../foo/trait.Trait.html" title="trait foo::Trait">Trait</a>&lt;'a&gt;,
}</pre></div><h2 id="fields" class="fields small-section-header">
                       Fields<a href="#fields" class="anchor"></a></h2><span id="structfield.some_func" class="structfield small-section-header"><a href="#structfield.some_func" class="anchor field"></a><code>some_func: {+for<'a>+} <a class="primitive" [-href="https://doc.rust-lang.org/nightly/std/primitive.fn.html">fn</a>'a(val:-]{+href="https://doc.rust-lang.org/nightly/std/primitive.fn.html">fn</a>(val:+} &amp;'a <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.i32.html">i32</a>) -&gt; <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.i32.html">i32</a></code></span>
```

camelid replied:
Fixed! https://github.com/rust-lang/rust/pull/79991#issuecomment-762553369

camelid commented on src/test/rustdoc/fn-type.rs:
```diff
@@ -0,0 +1,11 @@
+#![crate_name = "foo"]
+#![crate_type = "lib"]
+
+pub struct Foo<'a, T> {
+    pub generic: fn(val: &T) -> T,
+
+    pub lifetime: fn(val: &'a i32) -> i32,
+    pub hrtb_lifetime: for<'b, 'c> fn(one: &'b i32, two: &'c &'b i32) -> (&'b i32, &'c i32),
+}
```
I need to write `@has` checks for this.

camelid replied:
Done.

jyn514 commented:
> some_func: fn<'a>(val: &'a i32) -> i32

This is invalid syntax:
```
error: expected `(`, found `<`
 --> src/main.rs:1:30
  |
1 | fn higher_order(some_func: fn<'a>(val: &'a i32) -> &i32) {
  |                              ^ expected `(`
```
Is it simple to display `for<'a> fn` instead? If so, we should do that.

camelid commented:
Yes, I was planning on opening a follow-up PR; I figured we should do it in two stages so the rendered syntax is at least *plausible*. I can do it in this PR instead if you think that's better though.

jyn514 commented:
Yes, I think it would be better to fix it here if possible.

camelid commented:
I fixed it for bare functions, but I think it will be harder to fix for `dyn Trait`s. See the commit for more: cf5a620.

jyn514 commented on src/librustdoc/html/format.rs:
```diff
@@ -947,8 +950,14 @@ impl clean::FnRetTy {
-        comma_sep(self.generic_params.iter().map(|g| g.print()))
+    fn print_hrtb_bounds_with_space(&self) -> impl fmt::Display + '_ {
+        display_fn(move |f| {
+            if !self.generic_params.is_empty() {
```
This field is poorly named, I had to think why the only possible parameters could be HRTB: it's because `fn` is a function pointer and there's no place to *write* the generic parameter, `fn<T>(a: T)` is invalid syntax like we saw before. As a follow-up, not necessarily here, it would be nice to rename this to `hrtb` or something like that.

jyn514 replied:
Well, I guess it's named `generic_params` in the HIR :/ https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.BareFnTy.html this is kind of confusing but if it's used elsewhere I'd rather not mess with it.

camelid replied:
Marking as resolved since if we change it, it won't be in this PR.

jyn514 commented on src/librustdoc/html/render/mod.rs:
```diff
@@ -3048,6 +3048,7 @@ fn item_struct(w: &mut Buffer, cx: &Context, it: &clean::Item, s: &clean::Struct
                     ItemType::StructField,
                     field.name.as_ref().unwrap()
                 ));
+                debug!("LOOKIE: ty={:?}, printed={}", ty, ty.print());
```
:laughing: I think this might be better with a different prefix
```suggestion
                debug!("item_struct: ty={:?}, printed={}", ty, ty.print());
```

camelid replied:
Sorry, I thought I removed that debugging code! I use `LOOKIE` so it’s easy to search for in the logs :)

camelid replied:
I ended up just removing the debugging code since I didn't think it was useful enough. It's fairly straightforward to re-add if needed in the future.

jyn514 commented on src/librustdoc/html/format.rs:
```diff
@@ -947,8 +950,14 @@ impl clean::FnRetTy {
 impl clean::BareFunctionDecl {
-    fn print_generic_params(&self) -> impl fmt::Display + '_ {
-        comma_sep(self.generic_params.iter().map(|g| g.print()))
+    fn print_hrtb_bounds_with_space(&self) -> impl fmt::Display + '_ {
```
nit: 'higher ranked trait bounds bounds' is redundant.
```suggestion
    fn print_hrtb_with_space(&self) -> impl fmt::Display + '_ {
```

jyn514 commented:
> I fixed it for bare functions, but I think it will be harder to fix for dyn Traits. See the commit for more: cf5a620.

Ok, that seems reasonable. The new output for traits is at least *vaguely* right so this is definitely a step in the right direction :laughing:

jyn514 commented:
```
tidy error: /checkout/src/test/rustdoc/fn-type.rs: ignoring line length unnecessarily
```

jyn514 commented:
@camelid would this help with https://github.com/rust-lang/rust/issues/78482 ?

camelid commented:
> @camelid would this help with https://github.com/rust-lang/rust/issues/78482 ?

@jyn514 No, I don't think so. This just fixes the rendering on bare functions: the HRT bounds aren't rendered at all on trait bounds or trait objects.

camelid commented:
This PR is blocked on me figuring out why the `@has` checks are failing. Maybe @GuillaumeGomez knows? :P

GuillaumeGomez commented:
Can you provide the full diff provided by the `tidy` tool when running the tests too please? (You might need to install it, to update your nightly rustdoc and to rebase this PR though haha).

camelid commented:
Rebased.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1717807619) [(plain)](https://github.com/rust-lang/rust/commit/98cfbcc1b42eb491983b495c305ce02ede4fd356/checks/1717807619/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 29 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii

 finished in 0.082 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.35s

 finished in 2.432 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
failures:

---- [rustdoc] rustdoc/fn-type.rs stdout ----

error: htmldocck failed!
status: exit code: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/fn-type" "/checkout/src/test/rustdoc/fn-type.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
13: @has check failed
 `XPATH PATTERN` did not match
 // @has 'foo/index.html' '//span[@id="structfield.generic"]' 'generic: fn(val: &T) -> T'
Traceback (most recent call last):
  File "/checkout/src/etc/htmldocck.py", line 486, in <module>
    check(sys.argv[1], get_commands(sys.argv[2]))
  File "/checkout/src/etc/htmldocck.py", line 477, in check
    for c in commands:
  File "/checkout/src/etc/htmldocck.py", line 251, in get_commands
    args = shlex.split(args)
  File "/usr/lib/python3.6/shlex.py", line 305, in split
    return list(lex)
  File "/usr/lib/python3.6/shlex.py", line 295, in __next__
    token = self.get_token()
  File "/usr/lib/python3.6/shlex.py", line 105, in get_token
    raw = self.read_token()
  File "/usr/lib/python3.6/shlex.py", line 187, in read_token
    raise ValueError("No closing quotation")
ValueError: No closing quotation
------------------------------------------


info: generating a diff against nightly rustdoc
thread '[rustdoc] rustdoc/fn-type.rs' panicked at 'failed to exec `"rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/fn-type/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/fn-type.nightly" "/checkout/src/test/rustdoc/fn-type.rs"`', src/tools/compiletest/src/runtest.rs:1871:33

---- [rustdoc] rustdoc/for-lifetime.rs stdout ----


error: htmldocck failed!
status: exit code: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/for-lifetime" "/checkout/src/test/rustdoc/for-lifetime.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
11: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/index.html '//span[@id="structfield.some_func"]' "some_func: fn<'a>(val: &'a i32) -> i32"
12: @has check failed
 `XPATH PATTERN` did not match
 // @has foo/index.html '//span[@id="structfield.some_trait"]' "some_trait: dyn Trait<'a>"
Encountered 2 errors

------------------------------------------


info: generating a diff against nightly rustdoc
thread '[rustdoc] rustdoc/for-lifetime.rs' panicked at 'failed to exec `"rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/for-lifetime/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/for-lifetime.nightly" "/checkout/src/test/rustdoc/for-lifetime.rs"`', src/tools/compiletest/src/runtest.rs:1871:33

failures:
    [rustdoc] rustdoc/fn-type.rs
    [rustdoc] rustdoc/for-lifetime.rs
    [rustdoc] rustdoc/for-lifetime.rs

test result: FAILED. 411 passed; 2 failed; 2 ignored; 0 measured; 0 filtered out; finished in 32.17s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:18:33
```

</details>

camelid commented:
I fixed the test! It was a silly mistake on my part:

There were two issues:

  1. I was using the wrong filename: `foo/index.html` instead of
     `foo/struct.Foo.html`
  2. I was using the wrong syntax (I had changed it after creating the
     test): `fn<'a>` instead of `for<'a> fn`

The first one was the primary issue, though, because I couldn't
understand why the text I was matching against was the same as the text
in the `@has` check but the test was failing.

Glad I was able to figure it out!

camelid commented:
Should be ready for review! (Let me know before you `r+` and I can squash.)

GuillaumeGomez commented:
Looks good to me. Just waiting for @jyn514 now. ;)

bors commented:
:umbrella: The latest upstream changes (presumably #80987) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

camelid commented:
I'm going to squash before I rebase since the conflicts are non-trivial (and it
makes sense to squash anyway).

camelid commented:
Rebased!

@GuillaumeGomez Given that Joshua is taking a break for a bit, do you want to be
the reviewer for this? If not, I can wait until Joshua gets back, but I wanted
to check with you first :)

camelid commented on src/librustdoc/html/format.rs:
```diff
@@ -992,8 +995,14 @@ impl clean::FnRetTy {
+            if !self.generic_params.is_empty() {
+                write!(f, "for<{}> ", comma_sep(self.generic_params.iter().map(|g| g.print(cache))))
+            } else {
+                write!(f, "")
```
This should be just `Ok(())`. About to push that up.

GuillaumeGomez commented:
Looks good to me, thanks!

@bors: r=GuillaumeGomez,jyn514

bors commented:
:pushpin: Commit cd8dceef863378706bc27e0a3545c9251f02ec8b has been approved by `GuillaumeGomez,jyn514`

<!-- @bors r=GuillaumeGomez,jyn514 cd8dceef863378706bc27e0a3545c9251f02ec8b -->
<!-- homu: {"type":"Approved","sha":"cd8dceef863378706bc27e0a3545c9251f02ec8b","approver":"GuillaumeGomez,jyn514"} -->

hi-rustin commented:
> > @camelid would this help with #78482 ?
> 
> @jyn514 No, I don't think so. This just fixes the rendering on bare functions: the HRT bounds aren't rendered at all on trait bounds or trait objects.

@camelid  Do you have any suggestions for rendering it?

jyn514 commented:
@hi-rustin there are some instructions in https://github.com/rust-lang/rust/issues/78482. Please try to keep discussion in one place, this PR is not super relevant to 78482.
